qiskit/optimization/problems/has_quadratic_program.py

```diff
@@ -0,0 +1,45 @@
+# -*- coding: utf-8 -*-
+
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2019, 2020.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Interface for all objects that have a parent QuadraticProgram."""
+
+
+class HasQuadraticProgram:
+    """Abstract interface class for all objects that have a parent QuadraticProgram."""
+
+    def __init__(self, quadratic_program: "QuadraticProgram") -> None:
+        """ Initialize object with parent QuadraticProgram.
```

Comment:
Why is QuadraticProgram not imported here such that the typehints refer to the type itself rather than a string thereof.
The notion of `parent` here is confusing to me - the class has object as it parent implicitly not QuadraticProgram. This is a QuadraticProgramContainer that holds an instance of a QuadraticProgram. Can we think about the name of this class too - what I called it is a bit longer - HasQuadraticProgram seems more like some function's name that would return true or false as to whether it had one or not.

Reply:
The syntax that's frequently used in the stack and turns out to be very handy is
```
qp.linear_constraints[4].linear['x3'] = 5
```
In this example the linear constraint must resolve the string-name `'x3'` to an index. This is an information that only the quadratic program has, therefore the linear constraint must know to which program it is associated (hence parent).

We can probably design this a bit differently, the `HasQuadraticProgram` in most likely inspired from a Java-point of view. But since this is a change users will be unaffected of, let's do this in a refactor when we had enough time to come up with another design.